File: Algorithms/DivideConquer/BinarySearch.java

import java.util.Arrays;

public class BinarySeach {

    private static int binarySearch(int[] nums, int x) {

        int bottom = 0, top = nums.length - 1;

        while (bottom <= top) {

            int middle = (top + bottom) / 2;

            if (nums[middle] == x) {
                return middle;
            } else if (nums[middle] < x) {
                bottom = middle + 1;
            } else {
                top = middle - 1;
            }

        }

        return -1;

    }

    private static int binarySearch2(int[] nums, int x) {

        return recur(nums, 0, nums.length - 1, x);

    }

    private static int recur(int[] nums, int bottom, int top, int x) {

        if (top >= bottom) {

            int middle = (bottom + top) / 2;

            if (nums[middle] == x) {
                return middle;
            } else if (nums[middle] < x) {
                return recur(nums, middle + 1, top, x);
            }

            return recur(nums, bottom, middle - 1, x);

        }

        return -1;

    }

    public static void main(String[] args) {

        int[] nums = new int[]{3, 5, 657, 23, 5, 9, 2, 54, 22, 10, 14, 20, 42, 32, 15, 11};
        Arrays.sort(nums);
        System.out.println(Arrays.toString(nums));

        System.out.println(binarySearch(nums, 10));

        System.out.println("------");

        System.out.println(binarySearch(nums, 10));

    }

}

File: Algorithms/DivideConquer/BinarySearch.py

def binary_search(nums, x):

    bottom, top = 0, len(nums) - 1

    while bottom <= top:

        middle = (top + bottom) // 2

        if nums[middle] == x:
            return middle

        elif nums[middle] < x:
            bottom = middle + 1

        else:
            top = middle - 1

    return -1

def binary_search2(nums, x, bottom=None, top=None):

    if bottom == None or top == None:
        return binary_search2(nums, x, 0, len(nums) - 1)

    elif top >= bottom:

        middle = (top + bottom) // 2

        if nums[middle] == x:
            return middle

        elif nums[middle] < x:
            return binary_search2(nums, x, middle + 1, top)

        return binary_search2(nums, x, bottom, middle - 1)

    return -1

nums = sorted([3, 5, 657, 23, 5, 9, 2, 54, 22, 10, 14, 20, 42, 32, 15, 11])
print(binary_search(nums, 10))
print("------")
print(binary_search2(nums, 10))
